src/corppa/ocr/gvision_ocr.py

```diff
@@ -1,66 +1,293 @@
 
-        # save plain text output to local file;
-        # even if text is empty, create text file so we don't request again
-        with open(textfile, "w") as textfilehandle:
-            textfilehandle.write(response.full_text_annotation.text)
+            # Save plain text output to local file;
+            # even if text is empty, create text file so we don't request again
+            with open(textfile, "w") as textfilehandle:
+                textfilehandle.write(response.full_text_annotation.text)
 
-        # this generates a json string
-        json_response = vision.AnnotateImageResponse.to_json(response)
-        with open(jsonfile, "w") as jsonfilehandle:
-            jsonfilehandle.write(json_response)
+            # Save json response
+            json_response = vision.AnnotateImageResponse.to_json(response)
+            with open(jsonfile, "w") as jsonfilehandle:
+                jsonfilehandle.write(json_response)
 
-        if response.error.message:
-            raise Exception(
-                "{}\nFor more info on error messages, check: "
-                "https://cloud.google.com/apis/design/errors".format(
-                    response.error.message
+            if response.error.message:
+                raise Exception(
+                    f"{response.error.message}\n for more info on error messages, "
+                    "check: https://cloud.google.com/apis/design/errors"
                 )
+
+            # Update counter
+            ocr_count += 1
+            if show_progress:
+                progress_bar.update()
+
+            # Check if we should stop
+            if ocr_limit and ocr_count == ocr_limit:
+                # TODO: Is there a better structuring to avoid this break
+                break
+
+    if show_progress:
+        # Close progress bar
+        progress_bar.close()
+        if ocr_limit and ocr_count == ocr_limit:
+            print("Stopping early, OCR limit reached.", file=sys.stderr)
+        print(
+            f"{ocr_count:,} images OCR'd & {skip_count:,} images skipped.",
+            file=sys.stderr,
+        )
+
+    return {"ocr_count": ocr_count, "skip_count": skip_count}
+
+
+# Should this live somewhere in corppa.utils?
+def get_ppa_source(vol_id):
+    # Note that this is fairly brittle.
+    if vol_id.startswith("CW0") or vol_id.startswith("CB0"):
+        return "Gale"
+    else:
+        return "HathiTrust"
+
+
+def ocr_volumes(vol_ids, in_dir, out_dir, ext_list, ocr_limit=0, show_progress=True):
+    n_vols = len(vol_ids)
+    current_ocr_limit = ocr_limit
+    total_ocr_count = 0
+    total_skip_count = 0
+    for i, vol_id in enumerate(vol_ids):
+        vol_src = get_ppa_source(vol_id)
+        # Currently, this only supports Gale; skip other (i.e. HathiTrust) sources
+        if vol_src != "Gale":
+            print(
+                f"Warning: Skipping volume, since its source '{vol_src}' "
+                "is not yet supported.",
```

Comment:
Yes, this is an issue with the current HathiTrust directory structure on tigerdata. The issue is that right now the HathiTrust volume-level directory names include timestamps corresponding to the "version" timestamp.

In my mind this is a temporary issue, it'll work as soon as the volume-level directories correspond to the volumes "cleaned" id. One of the simplest fixes would be to move the version to a metadata file within the volume-level directory. That said, I think there's a larger question about how we want to handle the HathiTrust images since they aren't static like Gale.